includes/object-cache.php

```diff
@@ -0,0 +1,1160 @@
+	 * @param mixed $response
+	 * @return mixed
+	 */
+	protected function parse_redis_response( $response ) {
+		if ( is_bool( $response ) ) {
+			return $response;
+		}
+
+		if ( is_numeric( $response ) ) {
+			return $response;
+		}
+
+		if ( is_object( $response ) && method_exists( $response, 'getPayload' ) ) {
+			return $response->getPayload() === 'OK';
+		}
+
+		return false;
+	}
+
+	/**
+	 * Simple wrapper for saving object to the internal cache.
+	 *
+	 * @param   string $derived_key    Key to save value under.
+	 * @param   mixed  $value          Object value.
+	 */
+	public function add_to_internal_cache( $derived_key, $value ) {
+		$this->cache[ $derived_key ] = $value;
+	}
+
+	/**
+	 * Get a value specifically from the internal, run-time cache, not Redis.
+	 *
+	 * @param   int|string $key        Key value.
+	 * @param   int|string $group      Group that the value belongs to.
+	 *
+	 * @return  bool|mixed              Value on success; false on failure.
+	 */
+	public function get_from_internal_cache( $key, $group ) {
+		$derived_key = $this->build_key( $key, $group );
+
+		if ( isset( $this->cache[ $derived_key ] ) ) {
+			return $this->cache[ $derived_key ];
+		}
+
+		return false;
+	}
+
+	/**
+	 * In multisite, switch blog prefix when switching blogs
+	 *
+	 * @param int $_blog_id
+	 * @return bool
+	 */
+	public function switch_to_blog( $_blog_id ) {
+		if ( ! function_exists( 'is_multisite' ) || ! is_multisite() ) {
+			return false;
+		}
+
+		$this->blog_prefix = $_blog_id;
+
+		return true;
+	}
+
+	/**
+	 * Sets the list of global groups.
+	 *
+	 * @param array $groups List of groups that are global.
+	 */
+	public function add_global_groups( $groups ) {
+		$groups = (array) $groups;
+
+		if ( $this->redis_status() ) {
+			$this->global_groups = array_unique( array_merge( $this->global_groups, $groups ) );
+		} else {
+			$this->ignored_groups = array_unique( array_merge( $this->ignored_groups, $groups ) );
+		}
+	}
+
+	/**
+	 * Sets the list of groups not to be cached by Redis.
+	 *
+	 * @param array $groups List of groups that are to be ignored.
+	 */
+	public function add_non_persistent_groups( $groups ) {
+		$groups = (array) $groups;
+
+		$this->ignored_groups = array_unique( array_merge( $this->ignored_groups, $groups ) );
+	}
+
+	/**
+	 * Wrapper to validate the cache keys expiration value
+	 *
+	 * @param mixed $expiration Incomming expiration value (whatever it is)
+	 */
+	protected function validate_expiration( $expiration ) {
+		$expiration = ( is_array( $expiration ) || is_object( $expiration ) ? 0 : abs( intval( $expiration ) ) );
+
+		if ( $expiration === 0 && defined( 'WP_REDIS_MAXTTL' ) ) {
+			$expiration = intval( WP_REDIS_MAXTTL );
+		}
+
+		return $expiration;
+	}
+
+	/**
+	 * Unserialize value only if it was serialized.
+	 *
+	 * @param string $original Maybe unserialized original, if is needed.
+	 * @return mixed Unserialized data can be any type.
+	 */
+	protected function maybe_unserialize( $original ) {
```

Comment:
Compare against core functions, dig through commit log to find out why different.